doc/tpid/SonicTPIDSettingHLD1.md

```diff
@@ -0,0 +1,386 @@
+the CLI front-end and not allow the config go to the backend (Orchagent) in order to find out SAI is not capable and fail there. To achieve this
+CLI config precheck, SONiC during boot up will query SAI using SAI capability query API to check if the HW SKU can support TPID setting for
+PORT and LAG objects. This capability query result will be stored in the STATE\_DB so that the front-end CLI handler will be able to check and
+reject TPID setting on port/LAG if such capability is not supported by the SAI vendor running this HW platform. SONiC will be adding two
+entries in STATE\_DB under the existing "SWITCH\_CAPABILITY\|switch" table:
+
+*127.0.0.1:6379\[6\]\> hgetall \"SWITCH\_CAPABILITY\|switch\"*  
+*1) \"**PORT_TPID_CAPABLE**\"*  
+*2) \"**true**\"*  
+*3) \"**LAG_TPID_CAPABLE**\"*  
+*4) \"**false**\"*  
+*5) \"MIRROR\"*  
+*6) \"true\"*  
+*7) \"MIRRORV6\"*  
+*8) \"true\"*  
+*9) \"ACL\_ACTIONS\|INGRESS\"*  
+*10) \"PACKET\_ACTION,MIRROR\_INGRESS\_ACTION\"*  
+*11) \"ACL\_ACTIONS\|EGRESS\"*  
+*12) \"PACKET\_ACTION\"*  
+*13) \"ACL\_ACTION\|PACKET\_ACTION\"*  
+*14) \"DROP,FORWARD\"*  
+*127.0.0.1:6379\[6\]\>*  
+
+For the STATE\_DB shown above this switch/router is only capable of handling PORT TPID setting but not for LAG TPID setting.
+
+**TPID Configuration Flow**
+---------------------------
+
+Port/LAG TPID is configured by either CLI or Restful API talking to CFG manager, which check against State DB to ensure that the corresponding
+object TPID capability is "true" before proceed to update the corresponding port/LAG objects inside CFG DB. portsyncd once received
+the CFG Update notification will then push the config into PORT\_TABLE/LAG\_TABLE inside APP DB. If there is any change with
+PORT\_TABLE/LAG\_TABLE, its subscriber, PortsOrch, will write into ASIC DB and SYNCd will go through SAI to program the ASIC with the configured
+TPID value for the specified Port/LAG object.
+
+The following diagram summarizes how TPID configuration is handled within SONiC:
+
+![TPID Config Flow](TPID_ConfigFlow.png)
+
+
+**CFG Mgr TPID Setting Handling**
+---------------------------------
+
+As mentioned earlier, TPID config is only allowed if SAI capability check passes. Here is a sample CLI config attempt output on the config session
+if the SAI capability is not yet available:
+```
+admin@SONiC:~$ sudo config interface tpid Ethernet64 0x9200
+System not ready to accept TPID config. Please try again later.
+```
+
+Sample CLI config attempt output on the config session if the SAI capability for LAG is "false":
+```
+admin@SONiC:~$ sudo config interface tpid PortChannel0002 0x9200
+HW is not capable to support PortChannel TPID config.
+```
+
+In case a port is already configured as part of a LAG, TPID config directly to the LAG member port is not allowed and will fail as following:
+```
+admin@SONiC:~$ sudo config interface tpid Ethernet4 0x9200
+Ethernet4 is already member of PortChannel0002. Set TPID NOT allowed.
+```
+

```

Comment:
During the review there was one question raised by Ravi from Innovium in regard to the handling of adding a port to LAG when the port already has some non-default TPID and how should that be handled.
Here are 2 options:
1. After the port passes all current LAG membership add pre-check (if this is present today) we shall allow the add to proceed by first implicitly modify the port's TPID value to 0x8100 right before the Add LAG member continue to execute.
2. Reject the request due to Port TPID is NOT set at "default TPID" value. This forces the user to have to explicitly restore the port's TPID to 8100 first.  After the TPID is restored to default, the user can retry the add LAG member again and it should now able to proceed if TPID was the only blocking constraint check.

Option 1 has a potential race condition issue where if the implicit set Port TPID to default arrived later than the Add member to LAG.  So the code handling need to be aware of this race condition and only update the port's config DB, App DB to the default TPID value only and NOT program this TPID with SAI through syncd when such race condition is detected.

option 2 is the safest thing to do as it also draws attention to the user that issued the command.  Perhaps it was a typo where user specified the wrong interface... OR the error generated by this option can serves as a reminder to the user that the port being requested to add to the port channel may had some other usage intended.  It gives the user a chance to ensure it is desirable configuration that the user wants to perform.   The draw back of option 2 is that it prevented the user who know what he/she is doing and just want to quickly add the port to port-channel ignoring what its previous TPID maybe...

Given the two options above, I think we should go with option 2 for now...  If anyone else has comments over this, please raise a comment.
Thanks!

Reply:
Agree that option#2 is safer. 
And some more things to consider for these checks:  SAI allows a platform to specify the capability independently for a physical port and PortChannel. So we need to handle the theoretical possibility where the TPID is supported on the port and not on the PortChannel. And vice-versa.

Reply:
@rck-innovium
You are correct that the capability are independent and this HLD does not assume that the HW vendor is able to support both.  That is why in the State DB there are 2 separate attributes indicating whether TPID can be configured at PORT level and/or at LAG level.  
The requirement for support at LAG level actually placed more responsibility on the HW SAI vendor to be able to apply the TPID on each new joining member as well as reset to default TPID when a port is deleted from the LAG to become an individual port.   So my expectation is most vendor should have no problem to support at PORT level at first and later to support LAG but NOT vice versa as it make no sense to be able to support LAG but not able to support at port level... 
Rest assure this HLD when implemented will not make any assumptions for this.